### PR TITLE
Baked data for `icu_plurals`, `icu_timezone`

### PR DESCRIPTION
**Pull request by robertbastian**
#2743 

This adds data to subcrates (i.e. `<component>/data`). Users can provide custom baked data by setting the `ICU4X_DATA_DIR` environment variable during compilation. Crates will then fail to compile if the data doesn't contain the required keys.

The baked data uses `--locales modern`. I think that's a good default, if clients want other sets of languages they will have to use datagen.

While some of the generated files are big, the compressed tarballs for crates.io all fit comfortably within the 10MB limit.

```
64K     target/package/icu_calendar-1.2.0.crate
32K     target/package/icu_casemapping-0.7.2.crate
1.7M    target/package/icu_collator-1.2.0.crate
36K     target/package/icu_compactdecimal-0.2.0.crate
2.8M    target/package/icu_datetime-1.2.1.crate
20K     target/package/icu_decimal-1.2.0.crate
1.8M    target/package/icu_displaynames-0.10.0.crate
44K     target/package/icu_list-1.2.0.crate
60K     target/package/icu_locid_transform-1.2.1.crate
96K     target/package/icu_normalizer-1.2.0.crate
44K     target/package/icu_plurals-1.2.0.crate
316K    target/package/icu_properties-1.2.0.crate
392K    target/package/icu_relativetime-0.1.1.crate
7.4M    target/package/icu_segmenter-1.2.1.crate
20K     target/package/icu_timezone-1.2.0.crate
```

On the API side, I propose adding `try_new[_with_locale[_with_options]]` constructors. These provider-less constructors become the main documentation entry point, with constructors being listed in the order `*`, `_with_any_provider`, `*_with_buffer_provider`, `*_unstable`. The provider-constructors refer to the simple constructor's documentation. [Example](http://icu4x-pr-artifacts.storage.googleapis.com/gha/9a0950912a718e44def68ab2a71996bdfd0b299e/docs/icu_plurals/struct.PluralRules.html).

In locale-agnostic components (calendar, normalizer, locid_transform, properties, timezone, casemapping) we can create infallible constructors, which can usually be `const` (as these don't take arguments, they could also be `const` values [^1]). [Example](http://icu4x-pr-artifacts.storage.googleapis.com/gha/027f1808e2d2a523720070ff093c05d80c6bcae6/docs/icu_timezone/struct.MetazoneCalculator.html).

[^1]: My thinking here is to use `const fn` for constructors, and `const` values for what would be free functions, like in `icu_properties`.

### CODE REVIEW

File: components/timezone/src/metazone.rs
Comment (robertbastian) [inline components/timezone/src/metazone.rs:29]: can be `const` with #3467

File: components/timezone/src/metazone.rs
Comment (sffc) [inline components/timezone/src/metazone.rs:36]: Thought: maybe in cases where we could have a `pub fn new()`, we should instead export a static instance, which is probably better than people inlining calls to `new()` everywhere.

File: provider/core/src/constructors.rs
Comment (sffc) [inline provider/core/src/constructors.rs:117]: Nit: I do like moving these macros into this file, but I would prefer if you would do that in a standalone PR so that I can review the diffs better.

Comment (robertbastian) [reply provider/core/src/constructors.rs:117]: This is not a move, I'm adding a different version of the macro so we can migrate crates one by one.

File: components/timezone/src/metazone.rs
Comment (robertbastian) [reply components/timezone/src/metazone.rs:36]: I mention this in the pr description. I think for constructors `new` is more consistent with the provider constructors, and gets around the issue of naming the singleton. Performance is probably better with a const instead of a static, as that will be inlined.

For things like `load_emoji` I'm planning to have an `EMOJI` const.

File: components/timezone/src/metazone.rs
Comment (robertbastian) [reply components/timezone/src/metazone.rs]: actually `from_owned` can be `const`, so we can already do it.

File: components/plurals/build.rs
Comment (Manishearth) [inline components/plurals/build.rs:5]: thought: really not a fan of introducing more build scripts; they're a pain for non-Cargo users, and people in general do not like them.

Would prefer the crate based solution

File: components/plurals/data/macros.rs
Comment (Manishearth) [inline components/plurals/data/macros.rs:2]: issue: I would like us to prioritize fixing `rustfmt` on baked data; it's becoming really hard to figure out what's going on here.

I am very okay with having us use string interpolation instead

File: components/timezone/data.json
Comment (Manishearth) [inline components/timezone/data.json:1]: praise: i like this data.json plan

(worth moving into `data/` though? or calling it globaldata.json)

File: components/timezone/src/metazone.rs
Comment (Manishearth) [inline components/timezone/src/metazone.rs:23]: thought: Not quite sure we should do this.

**Review by Manishearth — DISMISSED** [review]

File: components/plurals/Cargo.toml
Comment (robertbastian) [reply components/plurals/Cargo.toml:57]: It's not really "global" though, it's just for this crate. If data is too general, I'd prefer "builtin-data".

File: components/plurals/build.rs
Comment (robertbastian) [reply components/plurals/build.rs:5]: In the issue we decided 

> globaldata can be replaced with a custom baked rust source using an env var.

There's no way to do defaulting env vars without build scripts.

File: components/timezone/data.json
Comment (robertbastian) [reply components/timezone/data.json:1]: `data/` gets overwritten

File: components/timezone/src/metazone.rs
Comment (robertbastian) [reply components/timezone/src/metazone.rs:23]: https://rust-lang.github.io/rust-clippy/v0.0.212/index.html#new_without_default

File: components/plurals/Cargo.toml
Comment (Manishearth) [reply components/plurals/Cargo.toml:57]: works for me

I think we've been calling it globaldata with global being real-world global

File: components/plurals/build.rs
Comment (robertbastian) [reply components/plurals/build.rs:5]: How does a separate crate solve this? It would only move the problem, you'd still have a build script there. It would also introduce another problems around semver, we'd need to use a `=` version range, but that's also frowned upon.

File: components/plurals/data/macros.rs
Comment (robertbastian) [reply components/plurals/data/macros.rs:2]: Does it matter? You can review the config that generates this.

File: components/plurals/build.rs
Comment (Manishearth) [reply components/plurals/build.rs:5]: Hmm. I was under the impression we were still considering doing crates. I think crate replacement works out far easier.

I would ideally like a non-build-script ICU4X to still _work_ with the data feature. To do this you'd have to use a `cfg` that's detected from the env var. Basically:

 - `ICU4X_PLURALS_REPLACE_DATA = ...` is the input
 - Build script detects it and sets `--cfg data-replacement` (see CPTBuilder's `icu4c_enable_renaming` for how this works)
 - Code uses `cfg(data-replacement)` to replace the relative path include with `include!(env!("ICU4X_PLURALS_REPLACE_DATA"))`

This means code using build scripts needs to set an env var _and_ a CFG, but that's not an undue burden IMO.

(If we're not doing crates, we don't really need macros anyway, right?)

File: components/timezone/src/metazone.rs
Comment (Manishearth) [reply components/timezone/src/metazone.rs:23]: aha, I see. hm. Makes sense I guess.

File: components/plurals/data/macros.rs
Comment (Manishearth) [reply components/plurals/data/macros.rs:2]: Yes, because this macro world is still new and I want to know what I'm reviewing. It's becoming hard to grasp the full end-to-end implications of these PRs; and relatedly when working on https://github.com/unicode-org/icu4x/pull/3480/ I was having a ton of trouble understanding what was going on in the regen.

I can potentially look into this tomorrow if it's not high priority for you.

File: components/plurals/build.rs
Comment (Manishearth) [reply components/plurals/build.rs:5]: Aha, I see.

Comment (Manishearth) [reply components/plurals/build.rs:5]: I hope we do not risk data duplication there. *might* be easier on the compiler to set up a fork-by-key provider for that

File: components/plurals/Cargo.toml
Comment (robertbastian) [reply components/plurals/Cargo.toml:57]: Then it should `moderndata` :)

File: components/plurals/build.rs
Comment (robertbastian) [reply components/plurals/build.rs:5]: We *could* actually do this without a build script if we had a `builtin-data` and a `custom-data` feature instead of feature + automatic cfg.

Comment (Manishearth) [reply components/plurals/build.rs:5]: Yeah dueling features are annoying

**Review by Manishearth — APPROVED** [review]

**Review by sffc — CHANGES_REQUESTED** [review]
Good work; need to align on open questions

File: components/plurals/build.rs
Comment (sffc) [reply components/plurals/build.rs:5]: Now thinking of it, it's really not correct that `_with_any_provider` and `_with_buffer_provider` call `_unstable` unconditionally; we see this problem in icu_locid_transform. It is more correct to call the current constructor type all the way down the stack, only converting to a DataProvider at the last minute when calling actual `load` functions. So, all four constructors should follow the same pattern of calling the same constructor of child classes and combining the results with `load` calls on the current crate's data structs.

File: components/plurals/src/lib.rs
Comment (robertbastian) [reply components/plurals/src/lib.rs:156]: I'm more and more of the opinion that anything with a provider should be considered a power-user feature. Google and Mozilla need it, but most users will not.

File: components/plurals/data.json
Comment (sffc) [reply components/plurals/data.json:1]: > Have we avoided it? Each ICU4X version is tied to a CLDR release

icu_datagen and icu_testdata are the only crates tied to a CLDR version.

We discussed some ideas for how to scheme the CLDR version in crates.io versions here: https://github.com/unicode-org/icu4x/issues/2540

> The globaldata feature is for clients who don't care about where the data comes from and just want ICU4X to work.

Acknowledged, but whether to use a single crate or separate crates for data is an (important) implementation detail visible  only in the lockfile. It shouldn't impact clients who don't care.

> Also keep in mind that we're releasing more frequently than CLDR, so we won't have to do data-only releases.

I hope that this won't forever be the case; once we are fully feature-complete, I want to slow down releases.

> If we put the data in separate crates, we have to stabilize the databake interface, which takes a lot of flexibility away from us.

We need to stabilize just the macros, right? We already export stable macros from databake. We're just adding some more.

Another reason to favor separate crates is that vendoring clients won't need to carry in all the big, unsafe databake files. They can disable the feature and avoid vendoring the data crates.

Do I think we need to answer this question before landing this PR? I think so, because this is a fairly fundamental decision that seems like it would be hard to decouple with future changes landing. I don't want to ship this to clients until we agree on the structure.

File: components/plurals/src/lib.rs
Comment (sffc) [reply components/plurals/src/lib.rs:156]: It's not great for sample code to require enabling a feature flag.

File: components/plurals/build.rs
Comment (robertbastian) [reply components/plurals/build.rs:5]: We seem to agree on the *Custom globaldata path* issue, and *Macros in crates* is not a discussion we need to have on this PR. I'm not calling macros from other crates.

File: components/plurals/data.json
Comment (robertbastian) [reply components/plurals/data.json:1]: > Do I think we need to answer this question before landing this PR? I think so, because this is a fairly fundamental decision that seems like it would be hard to decouple with future changes landing. I don't want to ship this to clients until we agree on the structure.

This issue is blocking 1.3 so merging does not risk shipping like this. I don't want to block the PR on this discussion, the bulk of changes is to the component APIs, whether we do `include!("../data/macros.rs")` or `use icu_globaldata::*` is not the main point.

File: components/plurals/src/lib.rs
Comment (robertbastian) [reply components/plurals/src/lib.rs:156]: And I assume you're opposed to making it a default feature?

**Review by sffc — COMMENTED** [review]
I'm not sure what you want me to re-review

File: components/plurals/data/Cargo.toml
Comment (sffc) [inline components/plurals/data/Cargo.toml:1]: oh I see, you implemented separate crates but distributed them inside the individual components instead of in a centralized location

Comment (robertbastian) [reply components/plurals/data/Cargo.toml:1]: Let's please not block on where in the repo these will live

File: components/plurals/src/lib.rs
Comment (robertbastian) [reply components/plurals/src/lib.rs:151]: Why does that matter? We can also pull in another crate by default.

File: components/plurals/data/build.rs
Comment (sffc) [inline components/plurals/data/build.rs:6]: Thought: You probably don't need to check for `CARGO_FEATURE_DATA`; the `cfg` should be ignored if the feature is disabled.

File: components/plurals/data/config.json
Comment (sffc) [inline components/plurals/data/config.json:10]: Need to discuss the CldrSet to use (doesn't need to block this PR)

File: tools/make/data.toml
Comment (sffc) [inline tools/make/data.toml:136]: Nit: I think the globaldata should be in its own job that could be run in the same CI as full-data, and it should have a diff-checker variant

File: components/plurals/data/Cargo.toml
Comment (robertbastian) [reply components/plurals/data/Cargo.toml:6]: Can we bikeshed versioning later?

File: tools/make/data.toml
Comment (robertbastian) [reply tools/make/data.toml:136]: Once we have data in all crates, this will be the same as the full-data check. We can then change the last line to `--locales basic moderate`, which will be super fast because the binary is already built.

Comment (robertbastian) [reply tools/make/data.toml:136]: Ah gotcha

**Review by sffc — APPROVED** [review]
Thanks for fixing my nitpicks :smiley: This is good to land now and we can continue bikeshedding the details

**Pull request merged**
